services/src/main/java/org/keycloak/protocol/oidc/grants/ciba/endpoints/BackchannelAuthenticationCallbackEndpoint.java

```diff
@@ -158,4 +180,52 @@ private void denyRequest(AccessToken authReqId, Status status) {
 
         store.deny(realm, authReqId.getId());
     }
+
+    protected void sendClientNotificationRequest(ClientModel client, CibaConfig cibaConfig, OAuth2DeviceCodeModel deviceModel) {
+        String clientNotificationEndpoint = cibaConfig.getBackchannelClientNotificationEndpoint(client);
+        if (clientNotificationEndpoint == null) {
+            event.error(Errors.INVALID_REQUEST);
+            throw new ErrorResponseException(OAuthErrorException.INVALID_REQUEST, "Client notification endpoint not set for the client with the ping mode",
+                    Response.Status.BAD_REQUEST);
+        }
+
+        logger.debugf("Sending request to client notification endpoint '%s' for the client '%s'", clientNotificationEndpoint, client.getClientId());
+
+        ClientNotificationEndpointRequest clientNotificationRequest = new ClientNotificationEndpointRequest();
+        clientNotificationRequest.setAuthReqId(deviceModel.getAuthReqId());
+
+        SimpleHttp simpleHttp = SimpleHttp.doPost(clientNotificationEndpoint, session)
```

Comment:
Recently we had a discussion around properly managing timeouts when calling downstream services. Ideally, we should be able to set a timeout specific for this context and sending the `can't.

It is not a blocker though, but if we could perhaps make the timeout configurable via CIBA config and change `SimpleHttp` to use this timeout to configure HTTP requests prior to execute them that would be nice.

Reply:
The timeouts are configurable already at the level of `DefaultHttpClientFactory` . This is for whole server, but IMO it can be sufficient for now?

As I see that there is not any other Keycloak functioality, which use timeouts specific per the functionality (including for example requests to Http Channel Provider, JWKS URL or all the other stuff where KEycloak sends some backend requests).

My vote is to add some more fine-grained timeouts support just if someone explicitly requests it for the client notification endpoint as each configuration option has some side-effect of the increased complexity...

Reply:
Yeah, there isn't. As I said not a blocker. It should be ok for now the default timeout.